include/rmm/device_scalar.hpp

```diff
@@ -39,49 +39,48 @@ class device_scalar {
    * @brief Construct a new `device_scalar`
    *
    * @param initial_value The initial value of the object in device memory
-   * @param stream_ Optional, stream on which to perform allocation and copy
-   * @param mr_ Optional, resource with which to allocate
+   * @param stream Optional, stream on which to perform allocation and copy
+   * @param mr Optional, resource with which to allocate
    *---------------------------------------------------------------------------**/
   explicit device_scalar(
-      T const &initial_value, cudaStream_t stream_ = 0,
-      rmm::mr::device_memory_resource *mr_ = rmm::mr::get_default_resource())
-      : buff{sizeof(T), stream_, mr_} {
-    auto status = cudaMemcpyAsync(buff.data(), &initial_value, sizeof(T),
-                                  cudaMemcpyDefault, buff.stream());
-
-    if (cudaSuccess != status) {
-      throw std::runtime_error{"Device memcpy failed."};
-    }
+      T const &initial_value, cudaStream_t stream = 0,
+      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
+      : buff{sizeof(T), stream, mr} {
+    
+    _memcpy(buff.data(), &initial_value, sizeof(T), stream);
   }
 
   /**---------------------------------------------------------------------------*
    * @brief Copies the value from device to host and returns the value.
    *
    * @return T The value of the scalar after synchronizing its stream
+   * @param stream CUDA stream on which to perform the copy
    *---------------------------------------------------------------------------**/
-  T value() const {
+  T value(cudaStream_t stream = 0) const {
     T host_value{};
-    auto status = cudaMemcpyAsync(&host_value, buff.data(), sizeof(T),
-                             cudaMemcpyDefault, buff.stream());
-    if (cudaSuccess != status) {
-      throw std::runtime_error{"Device memcpy failed."};
-    }
-    status = cudaStreamSynchronize(buff.stream());
-    if (cudaSuccess != status) {
-      throw std::runtime_error{"Stream sync failed."};
-    }
+    _memcpy(&host_value, buff.data(), sizeof(T), stream);
     return host_value;
   }
 
+  /**---------------------------------------------------------------------------*
+   * @brief Copies the value from host to device and synchronizes.
```

Comment:
We're coping from/to device/host pageable memory, and so synchronization is implicit.

https://docs.nvidia.com/cuda/cuda-driver-api/api-sync-behavior.html

I can add the synchonrization call explicitly, but I think that would cause the stream to sync twice.

https://github.com/rapidsai/rmm/pull/167#discussion_r340388661

Reply:
The fact that an async copy to non-paged memory is synchronous is different from saying that the function _synchronizes_.

Reply:
Fair. Reword to imply synchronization occurs, or remove synchronization comment entirely?

Reply:
```
/**---------------------------------------------------------------------------*
   * @brief Copies the value from device to host synchronously and returns the value.
   *
   * @return T The value of the scalar
   * @param stream CUDA stream on which to perform the copy
   *---------------------------------------------------------------------------**/
  T value(cudaStream_t stream = 0) const { ... };
```
```
/**---------------------------------------------------------------------------*
 * @brief Copies the value from host to device synchronously.
 *
 * @param host_value The host value which will be copied to device
 * @param stream CUDA stream on which to perform the copy
 *---------------------------------------------------------------------------**/
void set_value(T host_value, cudaStream_t stream = 0) {
```

Reply:
Either is fine.

However... It feels awfully strange to accept a stream in a function that is always synchronous.